reframe/frontend/statistics.py

```diff
@@ -1,89 +1,98 @@
+from itertools import chain
+
 import reframe.core.debug as debug
+import reframe.core.exceptions as exc
 
 
 class TestStats:
     """Stores test case statistics."""
 
-    def __init__(self, tasks=[]):
-        if not hasattr(tasks, '__iter__'):
-            raise TypeError('expected an iterable')
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = []
```

Comment:
This variable is also confusing. It's no more a list of tasks. It holds the runs per task. Honestly, I would prefer a dictionary. It would make the code much easier to read and would facilitate the implementation. I could understand the list of lists solution in a different language, but here I would definitely go for a dictionary.

```python
self._tasks_per_run = {}
```

Then you would initialise list of tasks simply as follows:

```python
self._task_per_run[self._current_run] = []
```

It can also be an `OrderedDict`, so as to keep the order of the keys (i.e., the runs).

Reply:
It does not hold the runs per task but the tasks ordered by (re-)try number. The (re-)try number is an integer (not a string); that's why I store it as a list of lists. 
The code has become much simpler compared to the previous implementation with dictionaries.